Zend/zend_alloc.c

```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
 
 static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 {
-	void *ptr = zend_mm_mmap(size);
-
+#ifdef _WIN32
+	void *ptr = NULL;
+	size_t offset;
+#ifndef _WIN64
+	/* 32-bit build has limited memory address space and partical deallocation is not supported,
+	 * try to reserve exact aligned size first, GH-9650 */
```

Comment:
The original code (and mmap code path) had `void *ptr = zend_mm_mmap(size);` and would return early if aligned and free it if that failed. I think it should continue doing that, and only do this as a fallback - was this still planned?

I assume it's faster to make 1 call to the allocator in the fast path than 3 (alloc + free + alloc at specific address)

Additionally, most of the time I'd expect allocated blocks to be allocated aligned to the 2MB boundary for long-running programs?

- e.g. if it's allocating blocks from the bottom up, then if the previous block was aligned to 2MB and of size 2MB, then the next free block of size 2MB will also usually be aligned

Reply:
intended and tested - `zend_mm_mmap(size)` returns almost always an unaligned virtual address, thus in the end, the original approach was doing much more allocation calls

for x86 we do it to address https://github.com/php/php-src/pull/9721#issuecomment-1277843700

Reply:
> zend_mm_mmap(size) returns almost always an unaligned virtual address (on windows), thus in the end, the original approach was doing much more allocation calls

Could you document that it's almost always unaligned on Windows?

Does that just apply to startup, or do you also mean that happens in subsequent allocations after the first in long-running php processes (In the original `PHP-8.0` branch, not this branch)

Reply:
`zend_mm_mmap(size)` is never tried to be allocated first for these reasons